src/Polysemy/Internal/TH/Effect.hs

```diff
@@ -58,152 +71,279 @@ makeSem = genFreer True
 -- function.
 --
 -- @
--- data Lang m a where
---   Output :: String -> Lang ()
+-- data Output o m a where
+--   Output :: o -> Output o m ()
 --
--- makeSem_ ''Lang
+-- makeSem_ ''Output
 --
--- -- | Output a string.
--- output :: Member Lang r
---        => String         -- ^ String to output.
+-- -- | Output the value \@o\@.
+-- output :: forall o r
+--        .  Member (Output o) r
+--        => o         -- ^ Value to output.
 --        -> Sem r ()  -- ^ No result.
 -- @
 --
--- Note that 'makeSem_' must be used /before/ the explicit type signatures.
+-- Because of limitations in Template Haskell, signatures have to follow some
+-- rules to work properly:
+--
+-- * 'makeSem_' must be used /before/ the explicit type signatures
+-- * signatures have to specify argument of 'Sem' representing union of
+-- effects as @r@ (e.g. @'Sem' r ()@)
+-- * all arguments in effect's type constructor that are type variables in
+-- action's signature have to use the same names as in type constructor; if
+-- they became more specific, they should use names from constructor's
```

Comment:
I don't necessarily understand what you mean by "more specific" here. Maybe elaborate?

Reply:
More specific = contains some concrete types; e.g. `()`, `Int` or `Maybe x` - in those cases we get concrete types (not really important) or new names from constructor's signature (like `x` above). Should I change the documentation?

Reply:
I suspect people will figure out the right thing to do. Maybe let's just chop out this bit about becoming more specific, and give the two examples.